smart-contracts/contracts/Unlock.sol

```diff
@@ -242,23 +242,47 @@ contract Unlock is
           // Get the value of 1 UDT (w/ 18 decimals) in ETH
           uint udtPrice = udtOracle.updateAndConsult(udt, 10 ** 18, weth);
 
-          // tokensToMint is either == to the gas cost times 1.25 to cover the 20% dev cut
-          uint tokensToMint = (estimatedGasForPurchase * tx.gasprice).mul(125 * 10 ** 18) / 100 / udtPrice;
-          // or tokensToMint is capped by percent growth
-          uint maxTokens = IMintableERC20(udt).totalSupply().mul(valueInETH) / 2 / grossNetworkProduct;
-          if(tokensToMint > maxTokens)
+          // tokensToDistribute is either == to the gas cost times 1.25 to cover the 20% dev cut
+          uint tokensToDistribute = (estimatedGasForPurchase * tx.gasprice).mul(125 * 10 ** 18) / 100 / udtPrice;
+
+          // or tokensToDistribute is capped by network GDP growth
+          uint gdpGrowth = valueInETH / grossNetworkProduct;
```

Comment:
You need to rework this approach so it only depends on whole numbers. If you don't need much precision, maybe basis points would be sufficient. e.g. 10000 * valueInETH / GNP. at the end you can then divide the 10000 back out.

Also impacts some of below.

Reply:
Ha! That explains why you had no "intermediate" number calculation in the existing code. I think precision is important because we're quickly going to talk about very small numbers (like a 1 DAI transaction on a 1Bn GDP is veery small but should still earn its "fair" share of UDT).

Reply:
Best way if you can rework the formulas so that all division is done once, at the very end -- e.g. `(a * b + c) / (x * y)`

Wolfgram is helpful for finding ways to reorder terms.

Reply:
Another possibility here is to look at how floating-point numbers are handled in the Locked.fyi hook.